Cloudmonitor: Refactor query builder

aangelisc opened the pull request:
This PR contains multiple fixes to ensure that builder values are refreshed according to the relevant property changes.

- All fields are refreshed when the `Service` is changed.
- All fields except `Filters` are refreshed when the `Metric name` is changed.
- The metric labels are refreshed when the panel time period is changed.
- Refactored `VisualMetricsQueryEditor` component and removed `Metrics` component.
- Updated `datasource` to expose `timeSrv`.

Fixes #58325

aangelisc committed:
Reset filters when service is changed

aangelisc committed:
Update to reset any query props

aangelisc committed:
Reset query properties on metric change

aangelisc committed:
Update tests

aangelisc committed:
Refresh labels on panel time update

aangelisc committed:
Merge branch 'main' into cloudmonitor-builder-updates

andresmgot commented on public/app/plugins/datasource/cloud-monitoring/components/VisualMetricQueryEditor.tsx:
```diff
@@ -38,12 +40,24 @@ function Editor({
   const [labels, setLabels] = useState<{ [k: string]: any }>({});
   const { projectName, groupBys, crossSeriesReducer } = query;
   const metricType = getMetricType(query);
+  const [timeState, setTime] = useState<{ from: string; to: string } | null>(null);
```
I think you can use `TimeOption` from `@grafana/data` here as the type.

andresmgot commented on public/app/plugins/datasource/cloud-monitoring/components/VisualMetricQueryEditor.tsx:
```diff
@@ -38,12 +40,24 @@ function Editor({
+  const useTime = (time: { from: string; to: string }) => {
+    if (timeState !== null && (timeState.from !== time.from || timeState.to !== time.to)) {
+      setTime({ ...timeSrv.time });
+    } else if (!timeState) {
+      setTime({ ...timeSrv.time });
+    }
```
what if rather than this `else` you do in the initialization:

```
  const timeSrv = getTimeSrv();
  const [timeState, setTime] = useState<TimeOption>(timeSrv.time);
```
?

andresmgot commented on public/app/plugins/datasource/cloud-monitoring/components/VisualMetricQueryEditor.tsx:
```diff
@@ -38,12 +40,24 @@ function Editor({
       datasource.getLabels(metricType, refId, projectName).then((labels) => setLabels(labels));
     }
-  }, [datasource, groupBys, metricType, projectName, refId, crossSeriesReducer]);
+  }, [datasource, groupBys, metricType, projectName, refId, crossSeriesReducer, timeState]);
```
this is a bit brittle, if you remove `timeState` here, nothing will tell you that you are breaking this. I see that `getLabels` is using `this.timeSrv.timeRange()`. Why don't you store the `TimeRange` in the state instead and pass it as an argument? That way you are sure that you are calling `getLabels` with the right range.

andresmgot replied:
not using `groupBys` and `crossSeriesReducer` already sounds like a bug and I don't know if that's on purpose :sweat_smile: 

> There isn't a need to pass TimeRange as an arg as the value used within the function would be the correct one

yes, but it feels more natural to calculate it here since this component is the one in charge of re-calling the function in case the TimeRange here. You are calculating the `time` here just to detect changes and recalculate the `range`. Isn't it simpler to just calculate the range and store that?

aangelisc replied:
Also a good point. I'm now considering doing a bigger refactor of the `VisualMetricQueryEditor` because I'm not entirely sure why this component is separate to the `Metrics` component when there isn't much in that component specifically, wdyt?

andresmgot replied:
yep, makes sense!

andresmgot reviewed (commented):
Noice :) I have some suggestions you may have already considered, let me know if you know more!

aangelisc commented on public/app/plugins/datasource/cloud-monitoring/components/Metrics.tsx:
```diff
@@ -109,6 +110,8 @@ export function Metrics(props: Props) {
         description: m.description,
       }));
+    // On service change reset all query values except the project name
+    Object.assign(query, { ...defaultTimeSeriesList(datasource), projectName: query.projectName });
```
Updating the service value triggers `onChange` as it causes the `onMetricTypeChange` function to be called 😊

andresmgot replied:
mmh, that would be relying on an indirect reference. `onMetricTypeChange` is getting the `query` object from `VisualMetricQueryEditor`, not from this component so it's a miracle if that works :sweat_smile:

I would move the definition of `onMetricTypeChange` to this component (since it's only used here) and explicitly build a full `query` object here that is used for the main `onChange` so it's explicit.

aangelisc committed:
Review

aangelisc committed:
Refactor VisualMetricsQueryEditor

- Move any Metrics functionality to VisualMetricsQueryEditor
- Update tests
- Expose timeSrv from datasource
- Update getLabels to make use of provided timeRange

joshhunt commented on .betterer.results:
```diff
@@ -5063,9 +5063,6 @@ exports[`better eslint`] = {
     "public/app/plugins/datasource/cloud-monitoring/components/MQLQueryEditor.tsx:5381": [
       [0, 0, 0, "Unexpected any. Specify a different type.", "0"]
     ],
-    "public/app/plugins/datasource/cloud-monitoring/components/Metrics.tsx:5381": [
-      [0, 0, 0, "Unexpected any. Specify a different type.", "0"]
-    ],
```
🥳

aangelisc committed:
Merge branch 'main' into cloudmonitor-builder-updates

andresmgot commented on public/app/plugins/datasource/cloud-monitoring/components/VisualMetricQueryEditor.test.tsx:
```diff
@@ -69,30 +221,66 @@ describe('VisualMetricQueryEditor', () => {
+    expect(await screen.queryByText('metric.test_groupby')).not.toBeInTheDocument();
+  });
+
+  it('updates labels on time range change', async () => {
```
:oof: tricky test :sweat_smile:

aangelisc replied:
🤣  I finally got it working and I'm very happy about this

andresmgot commented on public/app/plugins/datasource/cloud-monitoring/components/VisualMetricQueryEditor.test.tsx:
```diff
@@ -69,30 +221,66 @@ describe('VisualMetricQueryEditor', () => {
+
+    const service = await screen.findByLabelText('Service');
+    await openMenu(service);
+    await act(async () => {
```
I find that usually `act` is not needed, for example if you wait for the text 'Srv' to be there it should work the same way

aangelisc replied:
Without the `act` I have the error related to causing the state to be updated outside of `act` so I think it's safer to keep them in?

andresmgot commented on public/app/plugins/datasource/cloud-monitoring/components/VisualMetricQueryEditor.tsx:
```diff
@@ -13,125 +16,263 @@ import { Alignment } from './Alignment';
+  const [metrics, setMetrics] = useState<Array<SelectableValue<string>>>([]);
+  const [services, setServices] = useState<Array<SelectableValue<string>>>([]);
+  const [service, setService] = useState<string>('');
+  const [timeSrv, _setTimeSrv] = useState<TimeSrv>(datasource.timeSrv);
+  const [timeRange, setTimeRange] = useState<TimeRange>({ ...timeSrv.timeRange() });
```
why not
```suggestion
  const [timeRange, setTimeRange] = useState<TimeRange>({ ...datasource.timeSrv.timeRange() });
```
?

aangelisc replied:
Good shout!

andresmgot commented on public/app/plugins/datasource/cloud-monitoring/components/VisualMetricQueryEditor.tsx:
```diff
@@ -13,125 +16,263 @@ import { Alignment } from './Alignment';
+        label: m.displayName,
+        description: m.description,
+      }));
+    // On service change reset all query values except the project name
+    Object.assign(query, { ...defaultTimeSeriesList(datasource), projectName: query.projectName });
```
I don't think you need this assign, in  `if` condition below, `onMetricTypeChange` is doing the same, and in the `else` the query is not being updated anyway :shrug:

aangelisc replied:
This is partially true, I can change this to just set the `filters` value to `[]` as the assign in `onMetricTypeChange` conserves the value.

andresmgot reviewed (approved):
Thanks for the refactor :+1:

aangelisc committed:
Review
